src/library/Checkbox/Checkbox.js

```diff
@@ -127,4 +191,46 @@ export default class Checkbox extends PureComponent<Props> {
 
     return <Root {...rootProps} />;
   }
+
+  setInputRef = (node: HTMLInputElement) => {
+    const { inputRef } = this.props;
+    this.input = node;
+    inputRef && inputRef(node);
+  };
+
+  updateState = (indeterminate: boolean, checked: boolean) => {
+    this.indeterminate = indeterminate;
+    if (this.input) {
+      this.input.indeterminate = indeterminate;
+      this.input.checked = checked;
+    }
+  };
+
+  handleClick = (event: SyntheticInputEvent<HTMLInputElement>) => {
+    const { defaultIndeterminate, indeterminate, onChange } = this.props;
+
+    if (defaultIndeterminate || indeterminate) {
+      const nextIndeterminate = this.isControlled('indeterminate')
+        ? this.getIndeterminateValue()
+        : !this.getIndeterminateValue();
+
+      this.updateState(nextIndeterminate, nextIndeterminate);
+
+      if (!indeterminateClickTriggersChangeEvent) {
+        onChange && onChange(event);
+      }
+    }
+  };
+
+  isControlled = (prop: string) => {
+    return this.props.hasOwnProperty(prop);
+  };
+
+  getIndeterminateValue = () => {
```

Comment:
Since this returns a boolean & not a value, perhaps we just name it `isIndeterminate` or something similar?